### ISSUE TITLE
APIservice  triggers write: broken pipe of websocket

### ISSUE THREAD

**Issue by chinazj** [kubernetes/kubernetes#94990]
More debugging information :  https://github.com/kubernetes/kubernetes/issues/94963

**What happened**:

apiserver log

```
E0922 07:38:18.307950       1 available_controller.go:419] v1beta1.custom.metrics.k8s.io failed with: failing or missing response from https://10.0.7.105:443/apis/custom.metrics.k8s.io/v1beta1: Get https://10.0.7.105:443/apis/custom.metrics.k8s.io/v1beta1: write tcp 10.55.58.31:42324->10.0.7.105:443: write: broken pipe
E0922 07:38:18.308366       1 available_controller.go:419] v1beta1.metrics.k8s.io failed with: failing or missing response from https://10.0.7.105:443/apis/metrics.k8s.io/v1beta1: Get https://10.0.7.105:443/apis/metrics.k8s.io/v1beta1: write tcp 10.55.58.31:42324->10.0.7.105:443: write: broken pipe
E0922 07:38:19.882783       1 available_controller.go:419] v1beta1.webhook.provisioning.kommander.mesosphere.io failed with: failing or missing response from https://10.0.59.228:443/apis/webhook.provisioning.kommander.mesosphere.io/v1beta1: Get https://10.0.59.228:443/apis/webhook.provisioning.kommander.mesosphere.io/v1beta1: write tcp 10.55.58.31:56658->10.0.59.228:443: write: broken pipe
E0922 07:38:22.138323       1 available_controller.go:419] v1beta1.custom.metrics.k8s.io failed with: failing or missing response from https://10.0.7.105:443/apis/custom.metrics.k8s.io/v1beta1: Get https://10.0.7.105:443/apis/custom.metrics.k8s.io/v1beta1: write tcp 10.55.58.31:42324->10.0.7.105:443: write: broken pipe
E0922 07:38:22.138807       1 available_controller.go:419] v1beta1.webhook.provisioning.kommander.mesosphere.io failed with: failing or missing response from https://10.0.59.228:443/apis/webhook.provisioning.kommander.mesosphere.io/v1beta1: Get https://10.0.59.228:443/apis/webhook.provisioning.kommander.mesosphere.io/v1beta1: write tcp 10.55.58.31:56658->10.0.59.228:443: write: broken pipe
E0922 07:38:22.139418       1 available_controller.go:419] v1beta1.webhook.certmanager.k8s.io failed with: failing or missing response from https://10.0.28.36:443/apis/webhook.certmanager.k8s.io/v1beta1: Get https://10.0.28.36:443/apis/webhook.certmanager.k8s.io/v1beta1: write tcp 10.55.58.31:43788->10.0.28.36:443: write: broken pipe
E0922 07:38:22.139432       1 available_controller.go:419] v1beta1.metrics.k8s.io failed with: failing or missing response from https://10.0.7.105:443/apis/metrics.k8s.io/v1beta1: Get https://10.0.7.105:443/apis/metrics.k8s.io/v1beta1: write tcp 10.55.58.31:42324->10.0.7.105:443: write: broken pipe
E0922 07:38:27.317077       1 available_controller.go:419] v1beta1.webhook.certmanager.k8s.io failed with: failing or missing response from https://10.0.28.36:443/apis/webhook.certmanager.k8s.io/v1beta1: Get https://10.0.28.36:443/apis/webhook.certmanager.k8s.io/v1beta1: write tcp 10.55.58.31:43788->10.0.28.36:443: write: broken pipe
W0922 07:38:44.193505       1 request.go:157] Auditing failed of  request: encoding failed: /v1, Kind=DeleteOptions is unstructured and is not suitable for converting to "flagger.app/v1alpha3"
W0922 07:38:44.193504       1 request.go:157] Auditing failed of  request: encoding failed: /v1, Kind=DeleteOptions is unstructured and is not suitable for converting to "crd.projectcalico.org/v1"
W0922 07:38:44.199510       1 request.go:157] Auditing failed of  request: encoding failed: /v1, Kind=DeleteOptions is unstructured and is not suitable for converting to "velero.io/v1"
```

cert manager apiservice

```
[root@fort ~]# kubectl get apiservice  v1beta1.webhook.certmanager.k8s.io
NAME                                 SERVICE                                        AVAILABLE                      AGE
v1beta1.webhook.certmanager.k8s.io   cert-manager/cert-manager-kubeaddons-webhook   False (FailedDiscoveryCheck)   103d
```

kubectl describe

```
[root@fort ~]# kubectl describe apiservice  v1beta1.webhook.certmanager.k8s.io
Name:         v1beta1.webhook.certmanager.k8s.io
Namespace:
Labels:       app=webhook
              app.kubernetes.io/instance=cert-manager-kubeaddons
              app.kubernetes.io/managed-by=Tiller
              app.kubernetes.io/name=webhook
              helm.sh/chart=cert-manager-v0.10.1
Annotations:  certmanager.k8s.io/inject-ca-from-secret: cert-manager/cert-manager-kubeaddons-webhook-tls
API Version:  apiregistration.k8s.io/v1
Kind:         APIService
Metadata:
  Creation Timestamp:  2020-06-10T07:47:07Z
  Resource Version:    35215906
  Self Link:           /apis/apiregistration.k8s.io/v1/apiservices/v1beta1.webhook.certmanager.k8s.io
  UID:                 8487eb7a-a54c-4a83-bba6-6d146adca2b6
Spec:
  Ca Bundle:               LS0tLS1CRUdJTiBDRVJUSUZJQ0FURS0tLS0tCk1JSURVVENDQWptZ0F3SUJBZ0lRWFR0aVl2VEJ4Y2ZGUFlQSDhrLzRhekFOQmdrcWhraUc5dzBCQVFzRkFEQkEKTVJ3d0dnWURWUVFLRXhOalpYSjBMVzFoYm1GblpYSXVjM2x6ZEdWdE1TQXdIZ1lEVlFRREV4ZGpaWEowTFcxaApibUZuWlhJdWQyVmlhRzl2YXk1allUQWVGdzB5TURBMk1UQXdOelEzTVRKYUZ3MHlOVEEyTURrd056UTNNVEphCk1FQXhIREFhQmdOVkJBb1RFMk5sY25RdGJXRnVZV2RsY2k1emVYTjBaVzB4SURBZUJnTlZCQU1URjJObGNuUXQKYldGdVlXZGxjaTUzWldKb2IyOXJMbU5oTUlJQklqQU5CZ2txaGtpRzl3MEJBUUVGQUFPQ0FROEFNSUlCQ2dLQwpBUUVBMnJ0S0RVV0VvWDZOSVlVTFNyMXRPdGNtUFZBamY1dDZwNWd2TlZYcllXNFFmZTZhaG1nSEJjb0NQOFliCkNydFF4L3BXN3haTUIxNVRWbEwxUzJVQ1VTNjVwTjA2VE1QdC9obXRmcDI3bThSWEdGVmlpbHdma1FCak5KTjMKTFA5TXpneCtQRHRRSkdnRmtWdm5DUmFlVm5nY01tYjBpUkI0Y3p1emxQUDdpVUlJb3lFZVZSMjFzdnExN0s1cAp1VTVDZkFxKzNXcGV2NkJUMVpkWVNDbzJsN0VwVjN2Q1pVS0NQWWhYTElkQ1FaeUNyWHlTQ3dHVEUxZ1RSamRjCnlVUUpodHljWC9mYVZXUGdGSkpXbHBLM0lreDdrdWRiM0YyOTFrdjNBZSszay9JWlNqcncxTXovWTUwRFpQNUMKTytxQ2EyTkFkb2pKMm1BN01wWHRpT1ZoNHdJREFRQUJvMGN3UlRBT0JnTlZIUThCQWY4RUJBTUNBcVF3RHdZRApWUjBUQVFIL0JBVXdBd0VCL3pBaUJnTlZIUkVFR3pBWmdoZGpaWEowTFcxaGJtRm5aWEl1ZDJWaWFHOXZheTVqCllUQU5CZ2txaGtpRzl3MEJBUXNGQUFPQ0FRRUFaSWtJRFhmand3N0kyMVFFSUNJTVNRaVBrQ3U0djZGQzZMTGUKeFdraDBYOTlLYXZaOUYwR0VBSjBWYkdVUzB6b0VlRngrQyt4dTd4NkJZdC82VVRmNGtBaTlSTkF6RXFLbTVKRAp0cUpRdG0vaWZaelQzRmZyZlBZNGZUbWVScGhXa0NrSUJlV2xuQWYxNU1GaE4yZU1kNW45dnRETXExUUhGM2hQCjk1bjlWWUdBcTZrbTU3azJXdUJoQVlzbTd1RytJQklQYjQramdYQndiNURmWFBjV2hWTy81SVFobVZSNzlnMi8KYjd4UE9xVjRWajBzYXZIWjZBM3ZDaTBFejdGRURjWHJLMjRsdUxjaXU4RERmRlM1S2htSEZJMmkwakM2TmZLMQp5UDViV084cVY0MUNkRDZYQ040NnlmdnlDaklOSDU3Wm5SZ3NsZHY3ZGErcHNpMGtBdz09Ci0tLS0tRU5EIENFUlRJRklDQVRFLS0tLS0K
  Group:                   webhook.certmanager.k8s.io
  Group Priority Minimum:  1000
  Service:
    Name:            cert-manager-kubeaddons-webhook
    Namespace:       cert-manager
    Port:            443
  Version:           v1beta1
  Version Priority:  15
Status:
  Conditions:
    Last Transition Time:  2020-09-18T04:10:17Z
    Message:               failing or missing response from https://10.0.28.36:443/apis/webhook.certmanager.k8s.io/v1beta1: Get https://10.0.28.36:443/apis/webhook.certmanager.k8s.io/v1beta1: write tcp 10.55.58.31:43788->10.0.28.36:443: write: broken pipe
    Reason:                FailedDiscoveryCheck
    Status:                False
    Type:                  Available
Events:                    <none>
```

**What you expected to happen**:

apiserver does not have `write: broken pipe` problem

**How to reproduce it (as minimally and precisely as possible)**:

eth0 device is suddenly disconnected

```
ifdown eth0
ifup eth0
```

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`): 1.17.11
- Cloud provider or hardware configuration:  arm64
- OS (e.g: `cat /etc/os-release`):centos 7
- Kernel (e.g. `uname -a`): 3.10.0-1062
- Install tools:kubeadm
- Network plugin and version (if this is a network-related bug): calico
- Others:

**Comment by chinazj** [comment]
After the network card is normal. This error still occurs

**Comment by chinazj** [comment]
Check the documentation. It should be caused by the socket being closed, but writing characters to the socket

**Comment by yliaog** [comment]
what is your ValidatingWebhookConfiguration or MutatingWebhookConfiguration? what is the IP: 10.0.28.36, is it a cluster IP? Could you manually connect to 10.0.28.36:443 from the node that the k8s API server is deployed?

======================================================
    Message:               failing or missing response from https://10.0.28.36:443/apis/webhook.certmanager.k8s.io/v1beta1: Get https://10.0.28.36:443/apis/webhook.certmanager.k8s.io/v1beta1: write tcp 10.55.58.31:43788->10.0.28.36:443: write: broken pipe

**Comment by chinazj** [comment]
hi  @yliaog 
1.  it's ValidatingWebhookConfiguration.

```
➜  ~ kubectl get ValidatingWebhookConfiguration
NAME                                                             CREATED AT
auto-provisioning-validating                                     2020-09-11T07:50:20Z
cert-manager-kubeaddons-webhook                                  2020-09-22T07:02:36Z
dex-kubeaddons-dex-controller-validating-webhook-configuration   2020-06-10T07:50:44Z
dispatch-tekton-config.webhook.pipeline.tekton.dev               2020-08-27T07:35:34Z
dispatch-tekton-validation.webhook.pipeline.tekton.dev           2020-08-27T07:35:34Z
dstorageclass-validating-webhook-configuration                   2020-06-10T07:48:23Z
gatekeeper-kubeaddons                                            2020-06-10T07:48:30Z
prometheus-kubeaddons-prom-admission                             2020-08-27T07:31:59Z
```

```
➜  ~ kubectl get apiservice | grep webhook
v1beta1.webhook.certmanager.k8s.io                     cert-manager/cert-manager-kubeaddons-webhook                 True        6d3h
v1beta1.webhook.provisioning.kommander.mesosphere.io   konvoy/auto-provisioning-webhook                             True        17d
```

2. This is a bug caused by cert-manager and apiserver.  
The describe of the bug:  Apiserver `curl` the service of apiservice at regular intervals. After the error of `write: broken pipe`. When apiserver accesses the service again, it will remain unsuccessful.

Let's show this bug through the timeline

1.   APIService needs to be registered. (kubectl apply -f apiservice)

2.   apiserver successful `curl` cert-manager/cert-manager-kubeaddons-webhook. it may be websocket.

AVAILABLE is true
```
➜  ~ kubectl get apiservice | grep webhook
v1beta1.webhook.certmanager.k8s.io                     cert-manager/cert-manager-kubeaddons-webhook                 True        6d3h
v1beta1.webhook.provisioning.kommander.mesosphere.io   konvoy/auto-provisioning-webhook                             True        17d
```

3.  The client(cert-manager/cert-manager-kubeaddons-webhook) socket is closed and the server(apiserver) continues to write to the socket, This caused the error of write: broken pipe

AVAILABLE is FailedDiscoveryCheck
```
[root@fort ~]# kubectl get apiservice  v1beta1.webhook.certmanager.k8s.io
NAME                                 SERVICE                                        AVAILABLE                      AGE
v1beta1.webhook.certmanager.k8s.io   cert-manager/cert-manager-kubeaddons-webhook   False (FailedDiscoveryCheck)   103d
```

4. After a period of time, the socket of the client is opened. We can confirm that the client is normal through `curl`. But the server still has error: `write: broken pipe`

```
[root@fort ~]# kubectl exec -it tmp-shell /bin/bash
bash-5.0# curl -k --cert cert.crt --key cert.key https://10.0.28.36:443/apis/webhook.certmanager.k8s.io/v1beta1
{
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {

  },
  "status": "Failure",
  "message": "Unauthorized",
  "reason": "Unauthorized",
  "code": 401
}bash-5.0#
```

5.  We restarted this webhook (kubectl scale --replicas=0 ==> kubectl scale --replicas=0), but the error still exists, and the port numbers on both sides have not changed.

6.  After we restart the apiserver, it can connect normally. the error `write: broken pipe` disappear.

**Comment by chinazj** [comment]
https://github.com/kubernetes/kubernetes/issues/94963

**Comment by yliaog** [comment]
>>> 10.0.28.36:443 is cert-manager-webhook ( pod ip) . it's ok

i'm wondering why pod ip 10.0.28.36 is given directly to kube-apiserver, instead of a service ip?

>>> kubectl get apiservice  v1beta1.webhook.certmanager.k8s.io
>>> NAME                                 SERVICE                                        AVAILABLE                      AGE
>>> v1beta1.webhook.certmanager.k8s.io   cert-manager/cert-manager-kubeaddons-webhook   False (FailedDiscoveryCheck)   103d

instead of restarting the apiserver, could you try the following three steps?
1. "kubectl delete apiservice  v1beta1.webhook.certmanager.k8s.io"
2. after that, recreate the apiservice, 
3.  kubectl get apiservice  v1beta1.webhook.certmanager.k8s.io

**Comment by chinazj** [comment]
hi @yliaog 

1. This is my mistake，10.0.28.36:443 is  service ip.

```
➜  ~ kubectl get svc -A | grep webhook
cert-manager         cert-manager-kubeaddons-webhook                             ClusterIP      10.0.28.36    <none>         443/TCP                                                                                                                110d
dispatch             tekton-pipelines-webhook                                    ClusterIP      10.0.62.130   <none>         9090/TCP,8008/TCP,443/TCP                                                                                              49d
konvoy               auto-provisioning-webhook                                   ClusterIP      10.0.59.228   <none>         443/TCP                                                                                                                32d
kubeaddons           dex-kubeaddons-dex-controller-webhook-service               ClusterIP      10.0.60.86    <none>         443/TCP                                                                                                                110d
kubeaddons           dstorageclass-webhook-service                               ClusterIP      10.0.35.168   <none>         443/TCP                                                                                                                110d
kubeaddons           prometheus-webhook-dingtalk                                 ClusterIP      10.0.49.1     <none>         8060/TCP                                                                                                               68d
```

2. I tried to reinstall apiservice , but it was unsuccessful. The creation time is different. But I try to reproduce it

![image](https://user-images.githubusercontent.com/15140157/94508906-9e7e2c00-0245-11eb-85fe-c35e1d89fa52.png)

**Comment by chinazj** [comment]
Thanks for taking the time to read my question

**Comment by yliaog** [comment]
it is likely caused by the http2 golang client, the issue has been fixed in golang 1.15, but k8s 1.17 does not have it yet.

see 
https://github.com/kubernetes/client-go/issues/374#issuecomment-632457187
https://github.com/kubernetes/kubernetes/issues/80313#issuecomment-526451664
https://github.com/kubernetes/kubernetes/issues/75791#issuecomment-521719857

**Comment by chinazj** [comment]
it should be fixed #82090
